apps/vaporgui/PWidgetHLI.h

```diff
@@ -0,0 +1,132 @@
+CreateHLIBase( std::string, String );
+
+
+//
+// My attempt at implementing this using purely function templates and no defines. Below is pretty close.
+// It works but requires too much specilization code so I stopped working on it for now.
+//
+/*
+
+template <class Base, class Derived>
+void AssertInheritance()
+{
+    (void)static_cast<Base*>((Derived*)0);
+}
+
+template <class PW, typename T, class P, typename... Args>
+class PWidgetHLI : public PW {
+    typedef std::function<T(P*)> GetterType;
+    typedef std::function<void(P*, T)> SetterType;
+    
+    GetterType _getter;
+    SetterType _setter;
+public:
+    PWidgetHLI(std::string label, Args... args, GetterType getter, SetterType setter)
+    : PW("", args..., label) {
+        printf("PWidgetHLI Constructor (%s)\n", label.c_str());
+        AssertInheritance<PWidget, PW>();
+        AssertInheritance<VAPoR::ParamsBase, P>();
+        
+        this->_usingHLI = true;
+        
+        this->_getter = getter;
+        this->_setter = setter;
+    }
+    
+    virtual long getParamsLong() const override
+    {
+        static_assert(std::is_convertible<T, long>(), "");
+//        return this->_getterLong(this->getParams());
+        return this->_getter((P*)this->getParams());
+    }
+    
+    virtual void _setParamsLong(long v) override
+    {
+        static_assert(std::is_convertible<T, long>(), "");
+//        return this->_setterLong(this->getParams(), v);
+        this->_setter((P*)this->getParams(), v);
+    }
+    
+//    template <typename=typename std::enable_if<std::is_convertible<T, std::string>::type>>
+    virtual std::string getParamsString() const override
+    {
+        static_assert(std::is_convertible<T, std::string>(), "");
+//        return this->_getterLong(this->getParams());
+        return this->_getter((P*)this->getParams());
+    }
+    
+    virtual void _setParamsString(const std::string &v) override
+    {
+        static_assert(std::is_convertible<T, std::string>(), "");
+//        return this->_setterLong(this->getParams(), v);
+        this->_setter((P*)this->getParams(), v);
+    }
+};
+
+template <class PW, typename T, class P, typename... Args>
+class PWidgetHLI2 : public PWidgetHLI<PW, T, P, Args...> {
+    using PWidgetHLI<PW, T, P, Args...>::PWidgetHLI;
+};
+
+template <class PW, class P, typename... Args>
+class PWidgetHLI2<PW, std::string, P, Args...> : public PWidgetHLI<PW, std::string, P, Args...>
+{
+    using PWidgetHLI<PW, std::string, P, Args...>::PWidgetHLI;
+    virtual std::string getParamsString() const override
+    {
+    }
+};
+
+template <class PW, typename T, class P, typename... Args> typename std::enable_if<std::is_same<T, bool>::value, T>::type
+class PWidgetHLI2 : public PWidgetHLI<PW, T, P, Args...>
+{
+    virtual long getParamsLong() const override
+    {
+    }
+};
+
+//template <class PW, class P, typename... Args>
+//class PWidgetHLI2<PW, std::string, P, Args...> : public PWidgetHLI<PW, std::string, P, Args...> {
```

Comment:
Remove commented out code. Use ifdef if it is something we might want to preserve for later.

Reply:
Done